http/src/main/java/org/sapia/ubik/rmi/server/transport/http/JdkClientConnectionPool.java

```diff
@@ -15,45 +23,78 @@
  * sub-optimal implementation used only if the Jakarta HTTP client classes are
  * not in the classpath.
  * 
- * @author Yanick Duchesne
+ * @author yduchesne
  */
 public class JdkClientConnectionPool implements Connections {
-  private HttpAddress address;
-  private InternalPool pool = new InternalPool();
-
+  
+  private Category                      log    = Log.createCategory(getClass());
+  private HttpAddress                   address;
+  private Set<JdkRmiClientConnection>   active = Collections.newSetFromMap(new ConcurrentHashMap<>());
+  private InternalPool                  pool;
+  private JdkRmiClientConnectionFactory factory;
+  
+  /**
+   * @param address the address of the target server.
+   * @param factory the {@link JdkRmiClientConnectionFactory} to use.
+   * @param maxSize the pool's maximum size.
+   */
+  public JdkClientConnectionPool(HttpAddress address, JdkRmiClientConnectionFactory factory, int maxSize) {
+    this.address = address;
+    this.factory = factory;
+    this.pool = new InternalPool(maxSize);
+  }
+  
   /**
    * @param address
    *          the address of the target server.
+   * @param maxSize
+   *          the pool's maximum size.
    */
-  public JdkClientConnectionPool(HttpAddress address) {
-    this.address = address;
+  public JdkClientConnectionPool(HttpAddress address, int maxSize) {
+    this(address, () -> { 
+      JdkRmiClientConnection conn =  new JdkRmiClientConnection();
+      conn.setClock(RealtimeClock.getInstance());
+      return conn;
+    }, maxSize);
   }
-
+ 
+ 
   /**
    * @param transportType
    *          the "transport type" identifier.
    * @param serverUri
    *          the address of the target server.
+   * @param maxSize
+   *          the pool's maximum size.
    */
-  public JdkClientConnectionPool(String transportType, Uri serverUri) {
-    this(new HttpAddress(serverUri));
+  public JdkClientConnectionPool(String transportType, Uri serverUri, int maxSize) {
+    this(new HttpAddress(serverUri), maxSize);
   }
 
   @Override
   public RmiConnection acquire() throws RemoteException {
-    try {
-      return pool.acquire().setUp(address);
-    } catch (Exception e) {
-      if (e instanceof RemoteException) {
-        throw (RemoteException) e;
-      }
-
-      throw new RemoteException("Could acquire connection", e);
+    synchronized (pool) {
+      JdkRmiClientConnection connection = null;
+      try {
+        connection =  pool.acquire().setUp(address);
+        active.add((JdkRmiClientConnection) connection);
+        return connection;
+      } catch (Exception e) {
+        if (connection != null) {
```

Comment:
if the setup method above thows an exception, the connection variable will be null. The only safe way I see is to avoid method chaining when acquiring the connection

Reply:
Not sure I understand: if acquire() fails, an exception will be trhrown. That will happen before setUp() is called

Reply:
in `connection = pool.acquire().setUp(address);`, if the acquire() method succeeds but setUp() fail, the variable will never be assigned the connection acquired